wallet/src/actors/app/mod.rs

```diff
@@ -164,92 +190,105 @@ impl App {
                 let content = wallet::WalletContent::new(master_key, key_spec, vec![account]);
                 let info = wallet::WalletInfo {
                     id: id.clone(),
+                    name,
                     caption,
                 };
                 let wallet = wallet::Wallet::new(info, content);
 
                 slf.storage
                     .send(storage::CreateWallet(slf.db.clone(), wallet, password))
-                    .map_err(map_storage_failed_err)
+                    .map_err(app::Error::StorageFailed)
                     .map(move |_| id)
                     .into_actor(slf)
             });
 
         Box::new(fut)
     }
 
+    /// Unlock a wallet, that is, add its encryption/decryption key to the list of known keys so
+    /// further wallet operations can be performed.
     fn unlock_wallet(
         &mut self,
-        id: wallet::WalletId,
-        session_id: wallet::SessionId,
+        wallet_id: wallet::WalletId,
         password: ProtectedString,
-    ) -> ResponseActFuture<Self, (), Error> {
+    ) -> ResponseActFuture<Self, app::SessionId, app::Error> {
         // check if the wallet has already being unlocked by another session
-        match self.unlocked_wallets.get(&id).cloned() {
-            Some(mut owner_sessions) => {
-                log::debug!(
-                    "Wallet {} already unlocked. Appending {} to its list of active sessions.",
-                    &id,
-                    &session_id
-                );
-                owner_sessions.insert(id);
-                Box::new(fut::ok(()))
+        match self.wallet_keys.get(&wallet_id).cloned() {
+            Some(wallet_key) => {
+                let f = self
+                    .crypto
+                    .send(crypto::GenSessionId(wallet_key.clone()))
+                    .map_err(app::Error::CryptoFailed)
+                    .into_actor(self)
+                    .and_then(|id, slf, _ctx| {
+                        log::debug!("Wallet already unlocked by another session.");
```

Comment:
About this change, I'm not a big fan of logging sensible information (if someone gets the session id it will have access to that session's wallets), so I would prefer leaving it as: `log::debug!("Wallet `{}` was already unlocked another session", wallet_id);`, but what do you think?